test/functional/test_framework/test_node.py

```diff
@@ -51,8 +51,6 @@ def __init__(self, i, dirname, extra_args, rpchost, timewait, binary, stderr, mo
         self.extra_args = extra_args
         self.args = [self.binary, "-datadir=" + self.datadir, "-server", "-keypool=1", "-discover=0", "-rest", "-logtimemicros", "-debug", "-debugexclude=libevent", "-debugexclude=leveldb", "-mocktime=" + str(mocktime), "-uacomment=testnode%d" % i]
 
-        self.cli = TestNodeCLI(os.getenv("BITCOINCLI", "bitcoin-cli"), self.datadir)
```

Comment:
Thinking more about this, I think we should keep `TestNode.cli` to be the corresponding interface to `TestNode.rpc`. The logic to hack a different datadir can be placed in the test that checks for `-stdinrpcpass`. This makes the TestNode code easier to understand, as the datadir overwrite is rarely needed... Likely only once?

Reply:
I also think I prefer keeping `TestNode.cli` as a class (and making it callable as per https://github.com/jnewbery/bitcoin/commit/5bfd5e0d47f89f4c2e9becf266876c0c6cc62787), but I'm happy to be convinced the other way.

@MarcoFalke - I'm not sure I fully understand your comment though. As things currently stand with this PR, the interface for changing `datadir` is to add an argument to `TestNode.cli()`. How would you propose to make that different?

Reply:
In python every member is public, so changing the datadir is as easy as `self.nodes[0].cli.datadir = './new'  # Overwrite datadir, so we pass in rpcuser and password through stdin`

Reply:
yes, that works. I think I still prefer making `TestNodeCLI` callable and adding `datadir` as an argument.

Reply:
It is not like we need to pass in anything to `TestNodeCLI`. The `datadir` change is only needed for the test in this pull.

Reply:
Have you looked at https://github.com/jnewbery/bitcoin/commit/5bfd5e0d47f89f4c2e9becf266876c0c6cc62787 ?
Passing arguments to `TestNodeCLI` seems like a good way to pass `bitcoin-cli` command line arguments.

(I'm not arguing forcefully for the change in https://github.com/jnewbery/bitcoin/commit/5bfd5e0d47f89f4c2e9becf266876c0c6cc62787 and I'm happy to be shown a different way to do this)

Reply:
I did not.

This also help with selecting the rpcwallet, so making it callable is more useful than I thought.